torchtune/models/t5/_encoder.py

```diff
@@ -0,0 +1,304 @@
+                See: :class:`~torchtune.models.t5._encoder.T5EncoderRelativePositionBias`
+
+        Returns:
+            Tensor: output tensor with shape [bsz, seq_len, embed_dim]
+        """
+        bsz, seq_len, embed_dim = x.shape
+
+        # QKV projections
+        q = self.q_proj(x)
+        k = self.k_proj(x)
+        v = self.v_proj(x)
+
+        # [bsz, seq_len, embed_dim] -> [bsz, num_heads, seq_len, head_dim]
+        q = q.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+        k = k.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+        v = v.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+
+        # attention with relative position bias
+        attn_score = torch.matmul(q, k.transpose(-2, -1))
+        attn_score += rel_pos_bias
+        attn_weight = F.softmax(attn_score.float(), dim=-1).to(attn_score.dtype)
+        attn_out = torch.matmul(attn_weight, v)
+
+        # [bsz, num_heads, seq_len, head_dim] -> [bsz, seq_len, embed_dim]
+        attn_out = attn_out.transpose(1, 2).reshape(bsz, seq_len, embed_dim)
+
+        return self.output_proj(attn_out)
+
+
+class T5EncoderRelativePositionBias(nn.Module):
+    """
+    Computes binned birectional relative position bias for the T5 encoder.
+
+    It places relative positions into buckets and for each bucket, learns bias values for each attention head.
+
+    Args:
+        num_buckets (int): Number of discrete buckets to divide the relative positions into.
+        max_dist (int): Maximum distance for relative positions (distances beyond this are grouped into the last bucket)
+        num_heads (int): Number of attention heads in the transformer.
+        max_seq_len (int): Maximum sequence length (context length).
+    """
+
+    def __init__(
+        self, num_buckets: int, max_dist: int, num_heads: int, max_seq_len: int
+    ):
+        super().__init__()
+        self.max_seq_len = max_seq_len
+
+        # learnable mapping from bucket indices to bias values for each attention head
+        self.embedding = nn.Embedding(num_buckets, num_heads)
+
+        # fixed mapping from relative positions to bucket indices
+        self.register_buffer(
+            "relative_position_to_bucket",
+            _calc_birectional_rel_pos_to_bucket(num_buckets, max_dist, max_seq_len),
+            persistent=False,
+        )
+
+    def forward(self) -> Tensor:
+        """
+        Returns:
+            torch.Tensor: relative position bias tensor with shape [1, num_heads, max_seq_len, max_seq_len]
+        """
+        # convert bucket numbers to bias values for each attention head
+        x = self.embedding(self.relative_position_to_bucket)
+
+        # shape [max_seq_len, max_seq_len, num_heads] -> [1, num_heads, max_seq_len, max_seq_len]
+        return x.permute([2, 0, 1]).unsqueeze(0)
+
+
+def _calc_birectional_rel_pos_to_bucket(
```

Comment:
should I add more comments in this function explaining each operation? or is it fine to just leave it a bit opaque